Move at::{Refcounted,}MapAllocator to c10

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109881

`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

peterbell10 committed:
Move at::{Refcounted,}MapAllocator to c10

libshm.so depends on the torch library exclusively for
`at::RefcountedMapAllocator`, so it makes sense to move it to c10 along with the
other other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

peterbell10 commented on c10/core/MapAllocator.h:
```diff
@@ -0,0 +1,140 @@
+
+C10_API std::string NewProcessWideShmHandle();
+
+class C10_API MapAllocator {
```
These diffs are a mess, but it's mostly just a copy paste of the aten files with `aten` changed to `c10`. However, with the linter changes, it's too big of a diff for git to recognize as a simple rename. This is clearer in the commit-level diffs.

albanD reviewed (approved):
Ho very nice cleanup!

I'll verify that all build are good and then merge this.

albanD commented:
@pytorchbot rebase

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

albanD commented:
I can't import this so yolo

@pytorchbot merge -r

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

albanD commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge

albanD commented:
@pytorchbot revert -m "It does break buck build" -c nosignal

albanD commented on c10/core/build.bzl:
```diff
@@ -18,6 +18,27 @@ def define_targets(rules):
+            ":base",
+            "//c10/util:base",
+        ],
+        linkopts = ["-lrt"],
```
This is not actually a valid argument for buck rules.
Also this is not portable and rt should be pulled in as a deps.

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

albanD commented:
Is the ios failure relevant? The others are not for sure.

peterbell10 commented:
@pytorchbot rebase

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

peterbell10 commented:
iOS tests seems to be passing after the rebase so probably just a flake.

albanD commented:
Ok!
Feel free to merge this again once CI is green!

peterbell10 commented:
@pytorchbot merge

kit1980 commented:
@pytorchbot revert -m "breaking internal builds, undefined symbol: _ZN3c1022RefcountedMapAllocator6decrefEv" -c ghfirst

kit1980 commented:
@peterbell10 You probably need someone from Meta to help re-landing this.

peterbell10 commented on c10/BUCK.oss:
```diff
@@ -44,7 +45,12 @@ cxx_library(
-    preprocessor_flags = ['-DC10_BUILD_MAIN_LIB'],
+    preprocessor_flags = [
+        '-DC10_BUILD_MAIN_LIB',
+        '-DHAVE_MMAP=1',
```
```
undefined symbol: _ZN3c1022RefcountedMapAllocator6decrefEv"
```

I had the same error in the bazel build when `HAVE_MMAP` wasn't being defined. Is it possible there is another build definition that needs the flag added?

If so, does that mean we have 4 completely unrelated build definitions: CMake, bazel, buck (oss), buck (internal)?

kit1980 replied:
@peterbell10 unfortunately it's even more than 4: more than 1 internal buck.

But they are not completely unrelated: all of them are supposed to use build.bzl files (like https://github.com/pytorch/pytorch/blob/main/c10/build.bzl), but it's only implemented for some things so far.

For this particular case, likely https://github.com/pytorch/pytorch/blob/main/c10/ovrsource_defs.bzl needs to be updated. But someone from Meta should work with you on this.

peterbell10 committed:
Update on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

peterbell10 committed:
Fix rebase on "Move at::{Refcounted,}MapAllocator to c10"


`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

[ghstack-poisoned]

peterbell10 commented on c10/ovrsource_defs.bzl:
```diff
@@ -65,7 +68,9 @@ def define_c10_ovrsource(name, is_mobile):
-        deps = select({
+        deps = [
+            "//third-party/rt",
+        ] + select({
```
@kit1980 would you be able to help on the meta-internal side? I think this might fix it but not idea if it works.

kit1980 replied:
I can try but not today.

peterbell10 replied:
@kit1980 any updates on this?

lezcano replied:
ping @kit1980

kit1980 replied:
I'll look today.

kit1980 commented:
@pytorchbot rebase

kit1980 commented:
@peterbell10 May I ask you to fix the conflicts?
Or even better create a new non-ghstack PR - it will simplify things for me (it's a single PR stack anyway).

peterbell10 commented:
Okay, rebased and resubmitted as #116204
